commands/shell/tty_unix.go

```diff
@@ -0,0 +1,52 @@
+// +build linux darwin dragonfly freebsd netbsd openbsd
+
+package shell
+
+import (
+	"fmt"
+	"os"
+	"os/signal"
+	"syscall"
+
+	"github.com/burke/ttyutils"
+)
+
+// SetupRawTerminal will set terminal to RAW mode and return function to
+// restore normal terminal.
+func SetupRawTerminal(setSize func(cols, row uint16) error) func() {
+	// Switch terminal to raw mode
+	state, err := ttyutils.MakeTerminalRaw(os.Stdout.Fd())
+	if err != nil {
+		// If we can't do it just continue as we were
+		return func() {}
+	}
+
+	// Get TTY size
+	size, err := ttyutils.Winsize(os.Stdout)
+	if err == nil && setSize != nil {
+		setSize(size.Columns, size.Lines)
+	}
+
+	// Handle SIGWINCH signals (window resize signals from graphical terminals)
+	sigWinch := make(chan os.Signal, 1)
+	signal.Notify(sigWinch, syscall.SIGWINCH)
+	go func() {
+		for range sigWinch {
+			size, err := ttyutils.Winsize(os.Stdout)
+			if err == nil && setSize != nil {
+				setSize(size.Columns, size.Lines)
+			}
```

Comment:
I don't have much context here, but it feels like this code races. If the window size is modified between the calls to `WinSize` and `setSize`, it updates the size with invalid values. Not sure if it has practical consequences, or either if it has a non-racy solution.